include/Dialect/Openfhe/IR/OpenfheOps.td

```diff
@@ -90,15 +91,120 @@ def MulConstOp : Openfhe_Op<"mulconst",[
   let results = (outs RLWECiphertext:$output);
 }
 
+def MulAndRelinOp : Openfhe_Op<"mul_and_relin",[
+    Pure,
+    AllTypesMatch<["lhs", "rhs", "output"]>
+]> {
+  let arguments = (ins
+    Openfhe_CryptoContext:$cryptoContext,
+    RLWECiphertext:$lhs,
+    RLWECiphertext:$rhs
+  );
+  let results = (outs RLWECiphertext:$output);
+}
+
+def MulNoRelinOp : Openfhe_Op<"mul_no_relin",[
```

Comment:
Please extract this pattern as a base `class`, i.e. using the same organizational style as the CGGI Dialect: https://github.com/google/heir/blob/c21eb96dd5f33722682a20f88d6f844401fd2421/include/Dialect/CGGI/IR/CGGIOps.td#L25

It might look something like;

```
class Openfhe_BinaryOp<string mnemonic, list<Trait> traits = []>
  :  Openfhe_Op<mnemonic, traits # [
    Pure,
    AllTypesMatch<["lhs", "rhs", "output"]>,
]>{
 let arguments = (ins
    Openfhe_CryptoContext:$cryptoContext,
    RLWECiphertext:$lhs,
    RLWECiphertext:$rhs
  );
  let results = (outs RLWECiphertext:$output);
}
```

And each binary op can be defined then as

```
def MulNoRelinOp : Openfhe_BinaryOp<"mul_no_relin", [OpSpecificTrait1, ...]> { let summary = "..."; }
```

You could do a similar unary op base class as well. Up to you

Reply:
Applied!